# Homework 2 Regression Refresher

In class, we considered a very simply algorithm called gradient descent that we used to estimate the parameters of a simple linear model. But we left many considerations open. These include:
- Normalization 
    - We saw that the SSE is a very big number in our example, leading us to choose awkwardly small learning rates. Normaling the error may help us use more reasonably sized learning rates.
- When to stop
    - We can tell the algorithm to stop after a set number of iterations or, something smarter, tell it to stop when it begins to take very small step sizes
- Multiple solutions
    - It should be obvious that gradient descent, as an algorithm, doesn't do much "thinking". 
    - If there are multiple solutions, it will only find one
        

## Problem 1 - Use gradient descent to identify multiple minima

Regression algorithms ultimately search for minima. This can be where an error function is minimized (as we saw in class) or when a polynomial equation is minimized. Using the code from our lecture, adapt the code in order to indentify the roots of the following equation: 

$x^3 - 6x^2 +11x = 6$
  
- Look at homework 1 for how to tackle this problem.
- Since the polynomial will have more than 1 solution, be sure to consider multiple starting guesses. 
- Use the "roots" function from numpy to check your answer. Roots is especially designed for polynomials and should find all roots.

In [1]:
import numpy as np
from scipy.optimize import fsolve

def f(x):
    a=1
    b=-6
    c=11
    d=-6
    P = a*(x**3)+b*(x**2)+c*x+d
    return P
# Multiple Initial Guesses (tried different ways)

x = np.linspace(0.0, 5.0)
print (f(x))


[-6.         -4.93896251 -3.99649806 -3.16623176 -2.44178871 -1.81679402
 -1.2848728  -0.83965015 -0.47475117 -0.18380097  0.03957535  0.20175267
  0.3091059   0.36800993  0.38483965  0.36596996  0.31777576  0.24663193
  0.15891338  0.06099499 -0.04074833 -0.13994169 -0.2302102  -0.30517896
 -0.35847309 -0.38371767 -0.37453782 -0.32455864 -0.22740525 -0.07670273
  0.13392379  0.41084922  0.76044845  1.18909638  1.7031679   2.3090379
  3.01308128  3.82167294  4.74118777  5.77800066  6.93848651  8.22902022
  9.65597668 11.22573078 12.94465741 14.81913148 16.85552788 19.06022151
 21.43958725 24.        ]


#### From the results printed above, I decided to create an array from the values I understand were the solution.

In [2]:
#From the results printed above, I decided to create an array from the values I understand were the solution.
print(f(np.array([1.0,2.0,3.0])))


[0. 0. 0.]


### Using both methods we can see that the initial guesses: 1, 2, and 3, print out a value of 0 thus meaning these are the solutions.

In [3]:
# To check my answers
coeff = [1,-6,11,-6]
roots = np.roots(coeff)
print(roots)

[3. 2. 1.]


#### I didnt understand how to use the gradient decent code from the class to identify multiple minima. So I instead used Homework 1 as referance to solve the polynomial.

## Problem 2 - Linear regression can be used to analyze "groups"

In class, we used linear regression to fit a linear model to height and weight data. In this case, the measurements (x) and the outcomes (y) are continuous. Linear regression is also commonly used to analyze groups, wherein the measurements are not continous. For example, we may have measured the height of persons who did or did not eat asparagus when they were young. In this case, x is discrete - a collection of 1's and 0's. 

Let's use linear regression to analyze the miles per gallon achieved by select type of cars. Download the data from https://gist.github.com/noamross/e5d3e859aa0c794be10b. We want to determine if 8 cylinder vehicles get less gas mileage than 4 and 6 cylinder cars. Cyl = cylinders; MPG = miles per gallon on the table.

### a) Write the code to determine the average MPG for 8 cylinder vehicles vs 4/6 cyl vehicle.
Be sure the mean values are clearly identified or report them in a new markup cell below the code. Be sure that you are treating 4 and 6 cyl vehicles as one group.

In [4]:
%matplotlib inline 

import numpy as np
from io import StringIO
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns 

In [5]:
import csv

data = pd.read_csv('cars.csv')

#print (data)

### I first decided to determine the mean of mpg for each type of cylinder vehicle.
### The reason for this is to compare the mean values of the 8 cylinder vehicles obtain from here, to the ones obtain after grouping the 4 and 6 cylinder vehicles into one group.
### The mean mpg value of the 8 cylinder cars should be the same value for both times, this way I can check if my final values are correct.

In [6]:
avg_mpg = data.groupby('cyl')['mpg'].mean()

print(avg_mpg)

cyl
4    26.663636
6    19.742857
8    15.100000
Name: mpg, dtype: float64


### The mean mpg values of 8 cylinder vehicles is 15.1000.

In [7]:
data.columns = ['car_names', 'mpg','cyl', 'disp', 'hp', 'drat', 'wt', 'qsec', 'vs', 'am', 'gear', 'carb']
#data.head()
#data = data.drop(['car_names','disp', 'hp', 'drat', 'wt', 'qsec', 'vs', 'am', 'gear', 'carb'], axis=1)
data.head()


car_names   mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  \
0          Mazda RX4  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4   
1      Mazda RX4 Wag  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4   
2         Datsun 710  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4   
3     Hornet 4 Drive  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3   
4  Hornet Sportabout  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3   

   carb  
0     4  
1     4  
2     1  
3     1  
4     2

## I created two different group (0 and 1) based on the cylinder by applying a condition to each.

In [8]:
# Group 0 are for cylinder below 8 (but not including 8), therefore 4 and 6 cylinder:
data.loc[data.cyl < 8, 'cyl']=0

# Group 1 are for cylinder above or equal to 8:
data.loc[data.cyl >= 8, 'cyl'] = 1

# I tried two different way to determine the mean mpg to see if both worked the same 

## I tried two different way to determine the mean mpg to see if both worked the same 

In [9]:
# First option using this link as referance: https://www.askpython.com/python-modules/pandas/conditionally-grouping-values

In [10]:
# This "data.loc" lets me select the variables in the csv data file that I am interested in:
data_1 = data.loc[:, ["cyl", "mpg"]]

groupby = data_1.groupby("cyl", axis=0)
groupby.mean()

mpg
cyl           
0    23.972222
1    15.100000

In [11]:
# Second option using this link as referance: https://towardsdatascience.com/data-grouping-in-python-d64f1203f8d3

In [12]:
avg_mpg = data.groupby('cyl')['mpg'].mean()

print(avg_mpg)

#print (data)

cyl
0    23.972222
1    15.100000
Name: mpg, dtype: float64


## Final results for the mean mpg were for 4 and 6 cylinder cars 23.972222 and for 8 cylinder cars 15.100000.

### b) Analyze using linear regression
To analyze with linear regression, we have to create a linear model. Here, our y is the MPG. But x is not a continous variable - it's a group variable. 

- Let's create a "Group" column in our data, where a 1 indicates that a vehicle has a 8 cyl engine and a 0 to indicate a 4 or 6 cylinder vehicle. 
- Use the linear regression **mathematical code** (do not use packages; perform the matrix algrebra) from class to calculate the slope and intercept. Feel free to check your answer using on the packages demonstrated in class.
- Report the slope and intercept in a text box below your code. 
- In the text box, also compare the calculated slope and intercept to the values of the means calculated above. Do you notice any obvious relationships?



In [13]:
import csv

data = pd.read_csv('cars.csv')

# To create a new column in the data titles: "Group":

def new_group(row):
    if row['cyl'] < 8:
        return 0
    elif row['cyl'] == 8:
        return 1
data['Group']= data.apply(lambda row: new_group(row), axis=1)

# Here I named the columns of the initial data set and included the new group created:
data.columns = ['car_names', 'mpg','cyl', 'disp', 'hp', 'drat', 'wt', 'qsec', 'vs', 'am', 'gear', 'carb', 'Group']

# Since their is a lot of data I don't need, with this command "data.drop" I can eliminate them:
data = data.drop(['car_names','disp', 'hp', 'drat', 'wt', 'qsec', 'vs', 'am', 'gear', 'carb'], axis=1)

#databygroups = data.groupby('Group', axis=0)
data

mpg  cyl  Group
0   21.0    6      0
1   21.0    6      0
2   22.8    4      0
3   21.4    6      0
4   18.7    8      1
5   18.1    6      0
6   14.3    8      1
7   24.4    4      0
8   22.8    4      0
9   19.2    6      0
10  17.8    6      0
11  16.4    8      1
12  17.3    8      1
13  15.2    8      1
14  10.4    8      1
15  10.4    8      1
16  14.7    8      1
17  32.4    4      0
18  30.4    4      0
19  33.9    4      0
20  21.5    4      0
21  15.5    8      1
22  15.2    8      1
23  13.3    8      1
24  19.2    8      1
25  27.3    4      0
26  26.0    4      0
27  30.4    4      0
28  15.8    8      1
29  19.7    6      0
30  15.0    8      1
31  21.4    4      0

In [14]:
# Here the y would be the output so mpg:
y = data.loc[:, ["mpg"]]
#print(y)

# And x is the input so type of cylender car were I already identified 4 and 6 cylinders as 0 and 8 as 1.
x = data.loc[:, ["Group"]]
#print(x)

Y = np.c_[y]
#print(Y)
X = np.c_[np.ones(32), x]
#print(X)

from numpy.linalg import inv
tmp = inv(np.matmul(np.transpose(X),X))

INTERCEPT, SLOPE = tmp @ np.transpose(X) @ Y
print("Our intercept is ",INTERCEPT, "and our slope is", SLOPE)

Our intercept is  [23.97222222] and our slope is [-8.87222222]


### There is clear relationship between the intercept and slope values when compared to the previous means that makes sense.

### By assigning the x values based on the two groups (0 and 1) when x=0 the value of the intercept is the same as the mean value for 4 and 6 cylender cars and the slope is the difference of the two mean values.

## Github: https://github.com/cac407/Che_Math.git